src/ProductConstructionService/ProductConstructionService.Common/RedisMutex.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Logging;
+
+namespace ProductConstructionService.Common;
+
+public interface IRedisMutex
+{
+    public Task<T> DoWhenReady<T>(string mutexName, Func<Task<T>> action);
+}
+
+/// <summary>
+/// Mutex implemented using a redis cache meant for synchronization between repplicas
+/// </summary>
+public class RedisMutex : IRedisMutex
```

Comment:
Maybe this should be in the DependencyFlow project since it's only used there, but it is kind of universal

Reply:
I think here is fine